src/DotNetty.Codecs/Base64/Base64.cs

```diff
@@ -0,0 +1,205 @@
+﻿// Copyright (c) Microsoft. All rights reserved.
+// Licensed under the MIT license. See LICENSE file in the project root for full license information.
+
+namespace DotNetty.Codecs.Base64
+{
+    using System;
+    using DotNetty.Buffers;
+
+    public class Base64
+    {
+        const int MAX_LINE_LENGTH = 76;
+        const byte EQUALS_SIGN = (byte)'='; //pad
+        const byte NEW_LINE = (byte)'\n';
+        const sbyte WHITE_SPACE_ENC = -5; // Indicates white space in encoding
+        const sbyte EQUALS_SIGN_ENC = -1; // Indicates equals sign in encoding
+
+        public static IByteBuffer Encode(IByteBuffer src)
+        {
+            return Encode(src, Base64Dialect.STANDARD);
+        }
+
+        public static IByteBuffer Encode(IByteBuffer src, Base64Dialect dialect)
+        {
+            return Encode(src, src.ReaderIndex, src.ReadableBytes, dialect.breakLinesByDefault, dialect);
+        }
+
+        public static IByteBuffer Encode(IByteBuffer src, bool breakLines, Base64Dialect dialect)
+        {
+            return Encode(src, src.ReaderIndex, src.ReadableBytes, breakLines, dialect);
+        }
+
+        public static IByteBuffer Encode(IByteBuffer src, int offset, int length, bool breakLines, Base64Dialect dialect)
+        {
+            return Encode(src, offset, length, breakLines, dialect, src.Allocator);
+        }
+
+        public unsafe static IByteBuffer Encode(IByteBuffer src, int offset, int length, bool breakLines, Base64Dialect dialect, IByteBufferAllocator alloctor)
+        {
+            if (src == null)
+            {
+                throw new NullReferenceException("src is null");
+            }
+            if (dialect.alphabet == null)
+            {
+                throw new NullReferenceException("alphabet is null");
+            }
+            if (dialect.alphabet.Length != 64)
+            {
+                throw new Exception(String.Format("alphabet.Length:{0}, not 64!", dialect.alphabet.Length));
+            }
+            if (length <= 0) return Unpooled.Empty;
+
+            int outLength = length / 3 * 4;
+            outLength += length % 3 > 0 ? 4 : 0;
+            outLength += breakLines ? outLength / MAX_LINE_LENGTH : 0;
+            IByteBuffer dest = alloctor.Buffer(outLength).WithOrder(src.Order);
+            //avoid unnecessary range checking
+            fixed (byte* alphabet = dialect.alphabet, srcArray = src.Array, destArray = dest.Array)
+            {
+                int i = 0;
+                int j = 0;
+                int charCount = 0;
+                int calcLength = offset + length - length % 3;
```

Comment:
I just don't trust the compiler with optimizations I guess. Can you guarantee that compiler will optimize introduce variable itself for "length % 3" to avoid calculating it multiple times? If so, then it's ok to leave it as is.